gridfs.go

```diff
@@ -0,0 +1,601 @@
+package pail
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"regexp"
+	"time"
+
+	"github.com/mongodb/grip"
+	"github.com/mongodb/grip/message"
+	"github.com/pkg/errors"
+	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/bson/primitive"
+	"go.mongodb.org/mongo-driver/mongo"
+	"go.mongodb.org/mongo-driver/mongo/gridfs"
+	"go.mongodb.org/mongo-driver/mongo/options"
+)
+
+// GridFSOptions support the use and creation of GridFS backed buckets.
+type GridFSOptions struct {
+	Name         string
+	Prefix       string
+	Database     string
+	MongoDBURI   string
+	DryRun       bool
+	DeleteOnSync bool
+	DeleteOnPush bool
+	DeleteOnPull bool
+	Verbose      bool
+}
+
+func (o *GridFSOptions) validate() error {
+	if (o.DeleteOnPush != o.DeleteOnPull) && o.DeleteOnSync {
+		return errors.New("ambiguous delete on sync options set")
+	}
+
+	return nil
+}
+
+type gridfsBucket struct {
+	opts   GridFSOptions
+	client *mongo.Client
+}
+
+func (b *gridfsBucket) normalizeKey(key string) string { return b.Join(b.opts.Prefix, key) }
+
+func (b *gridfsBucket) denormalizeKey(key string) string {
+	if b.opts.Prefix != "" && len(key) > len(b.opts.Prefix)+1 {
+		key = key[len(b.opts.Prefix)+1:]
+	}
+	return key
+}
+
+// NewGridFSBucketWithClient returns a new bucket backed by GridFS with the
+// existing Mongo client and given options.
+func NewGridFSBucketWithClient(ctx context.Context, client *mongo.Client, opts GridFSOptions) (Bucket, error) {
+	if client == nil {
```

Comment:
Given that this variant of the constructor is meant to pass a client, I feel like this should just error if the client is nil, rather than magic up a bucket with a new client.